src/compiler/checker.ts

```diff
@@ -3422,12 +3471,18 @@ module ts {
                         else {
                             type = getDeclaredTypeOfSymbol(symbol);
                             if (type.flags & (TypeFlags.Class | TypeFlags.Interface) && type.flags & TypeFlags.Reference) {
-                                let typeParameters = (<InterfaceType>type).typeParameters;
-                                if (node.typeArguments && node.typeArguments.length === typeParameters.length) {
-                                    type = createTypeReference(<GenericType>type, map(node.typeArguments, getTypeFromTypeNode));
+                                // In a type reference, the outer type parameters of the referenced class or interface are automatically
+                                // supplied as type arguments and the type reference only specifies arguments for the local type parameters

```

Comment:
That is true only for written type references. Type references created by instantiations will indeed end up having the type parameters mapped to different types. For example:
```typescript
function f<T>() {
    class C {
        x: T;
    }
    var c: C;  // Really a C<T>
    c = new C();
    return c;
}
var c = f<string>();  // Type of c is C<string>
var s = c.x;          // string
```